### DISCUSSION TITLE
Guided Prompting

### DISCUSSION THREAD

**Discussion by shaneholder** [Azure/azure-dev#3120] [General]
Hello, I've been working on a similar project that is internal to my company using yeoman but considering switching our templates to use azd instead.  Our current system will prompt users for some additional Azure resources like vnets and keyvaults.  We build out workloads using the CAF vending machine concept and the initial subscription layout will create a few resources that we consider necessary for all projects including a couple of vnets, one peered to a corporate vwan.  I would like to be able to prompt a user for the vnet/subnet that a resource should be placed into, is that possible or on the roadmap?

**Comment by lechnerc77** [comment]
Hi @shaneholder , did you take a look at the option of using `hooks` to implement this requirement? You find the official documentation here: <https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/azd-extensibility>.

This should allow you in a scipted and interactive approach to query the parameters and then use them in azd.

**Comment by shaneholder** [comment]
Hi. Christian,

No I had not considered that.  I did look at the hooks and see their value
but did not think of them as a solution to this problem.  I will re
evaluate that and see what I can put together.  However just at first
thought it seems like a feature that should be more built into the
product.  Building a navigable and searchable list of choices with names
and values could end up being more complicated than necessary but it is an
interesting option.

Thanks for the idea.

Regards,
Shane

**Comment by vhvb1989** [comment]
@shaneholder , if use set an input parameter without a default value, azd will prompt the user to enter a value for it.

See this example: https://github.com/Azure-Samples/azure-search-openai-demo/blob/6cc78369938e463085ae777bc83a82adc27e40fe/infra/main.bicep#L54

```bicep
@description('Location for the OpenAI resource group')
@allowed(['canadaeast', 'eastus', 'eastus2', 'francecentral', 'switzerlandnorth', 'uksouth', 'japaneast', 'northcentralus', 'australiaeast', 'swedencentral'])
@metadata({
  azd: {
    type: 'location'
  }
})
param openAiResourceGroupLocation string
```

The parameter `openAiResourceGroupLocation` has no default value, so `azd` will automatically display a prompt before provisioning the infrastructure, asking for the value.

azd supports the `azd/type` object as part of the parameter metadata to make azd to use an specific prompt type.  The only supported type until now is `location`, which would show the user a list of locations interactively.

If the azd/type metadata is not provided, azd will just use a free-input-text as input.

Can you describe how, in your use-case, a user would choose a `vnet/subnet` ?  
Would the user already know the name to use?  or do the user need to pull the valid names from somewhere?
Based on your feedback, we could create a feature-request to support a unique prompt, similar to the `location` prompt, where you can set the valid `vnet/subnet`

**Reply by shaneholder** [reply]
To the rest of your question, no the user will likely not know the name of the vnet/subnet.  We use az cli with the --query to pull name and value to display to the user based on the subscription they selected.  Then if they're asking to add a VM we know the subnet to create the VM in, we do a similar process for keyvault for VM access.  It's possible that there are multiple KV's in a Subscription (we suggest that users use a separate keyvault for things like SSL Certificates for their application gateways) so we prompt the user for which KV they want to put the default VM credentials into.

**Reply by shaneholder** [reply]
Here's an example az command we execute.
```
az network vnet list  --query "[].{name:name, value: {name:name, location:location, range:addressSpace, rg:resourceGroup, subnets:subnets[].{name:name, value:name}}}"
```

This allows us to pull several bits of information.  Once the user selects a vnet then we prompt for the subnet which gives us cidr ranges to use in various places. 

Hope that helps some.

**Reply by vhvb1989** [reply]
I created a new feature-request for you: https://github.com/Azure/azure-dev/issues/3129

azd should help you to pick a vnet, instead of using `az`.

**Comment by shaneholder** [comment]
That is interesting as well, does that work for terraform?

Thanks,
Shane

**Reply by vhvb1989** [reply]
No, the prompting (and azd metadata) is unique to bicep.
For terraform, you would get the terraform-cli default behavior for input parameters.
